### PR TITLE
Add fetchRevisions hook to support revision listing

### PR DESCRIPTION
**Pull request by jheth**
I used ember-cli-deploy-s3-index as an example for implementing `fetchRevisions`

`fetchRevisions` has been hooked into `activate` to ensure a valid revision was specified

### CODE REVIEW

**Comment by jeffjewiss** [comment]
Thanks for this! I should have a chance to review it over the next few days.

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (jeffjewiss) [inline .vscode/launch.json]: Could you please remove this file, or if you think it should be added to the project, could you please move it to another PR?

File: index.js [hunk 2]
```diff
@@ -85,20 +85,103 @@ module.exports = {
...
+        return this._list(context)
+          .then(function(revisions) {
+            return {
+              revisions: revisions
+            };
+          });
+      },
+
+      _list: function(/* context */) {
+        let bucket      = this.readConfig('bucket');
```
Comment (jeffjewiss) [inline index.js]: Since we're not aligning the `=` could you please collapse the whitespace on this line and the `let`s below so that there's only one space?

File: index.js [hunk 3]
```diff
@@ -85,20 +85,103 @@ module.exports = {
...
+        let deployArchive      = this.readConfig('deployArchive');
+        let revisionPrefix = deployArchive + "-";
```
Comment (jeffjewiss) [inline index.js]: Could these be template strings instead?
```js
let revisionPrefix = `${deployArchive}-`;
```

File: index.js [hunk 4]
```diff
@@ -85,20 +85,103 @@ module.exports = {
...
+
+        return RSVP.hash({
+          revisions: this.listObjects(this.s3, { Bucket: bucket, Prefix: revisionPrefix }),
+          current: this.getObject(this.s3, { Bucket: bucket, Key: indexKey }),
+        })
+        .then(function(data) {
+          let activeRevision = '';
+          if (data.current) {
+            let objectData = data.current.Body.toString('utf-8');
+            if (objectData[0] === '{') {
```
Comment (jeffjewiss) [inline index.js:142]: Do you think it would be helpful to have `else` statements that throw errors so that the parsing of the data doesn't fail silently or was this not an issue in your testing?

File: index.js [hunk 5]
```diff
@@ -85,20 +85,103 @@ module.exports = {
...
+            let revision = d.Key.substring(revisionPrefix.length, d.Key.lastIndexOf('.'));
+            let active = data.current && revision === activeRevision;
+            return { revision: revision, timestamp: d.LastModified, active: active, deployer: 'fastboot-s3' };
+          });
+
+          if (results.length && results[0].revision == 'info') {
+            results.shift();
+          }
+          return results;
+        });
```
Comment (jeffjewiss) [inline index.js]: Could a `catch` be added here to handle data parsing errors similar to how it's handled on other methods? 
```js
.catch(_this._errorMessage.bind(_this));
```

File: index.js [hunk 6]
```diff
@@ -85,20 +85,103 @@ module.exports = {
...
+
+          let results = data.revisions.Contents.sort(function(a, b) {
+            return new Date(b.LastModified) - new Date(a.LastModified);
+          }).map(function(d) {
+            let revision = d.Key.substring(revisionPrefix.length, d.Key.lastIndexOf('.'));
+            let active = data.current && revision === activeRevision;
+            return { revision: revision, timestamp: d.LastModified, active: active, deployer: 'fastboot-s3' };
+          });
+
+          if (results.length && results[0].revision == 'info') {
```
Comment (jeffjewiss) [inline index.js]: Should this be a `===`?

File: index.js [hunk 7]
```diff
@@ -85,20 +85,103 @@ module.exports = {
...
+               })
+               .catch(_this._errorMessage.bind(_this));
+
+           } else {
+             return RSVP.reject("REVISION NOT FOUND!"); // see how we should handle a pipeline failure
+           }
+         });
+      },
+
+      fetchRevisions: function(context) {
```
Comment (jeffjewiss) [inline index.js]: Could you please update the new methods to follow the `fetchRevisions(context) {` style?

File: index.js (hunk 3 above)
Comment (jeffjewiss) [inline index.js]: This causes an error if a `prefix` is used (ie. putting the files in a directory in the bucket). Could you please add the prefix value to the `revisionPrefix` if it exists? This was done elsewhere with a ternary.

File: index.js [hunk 8]
```diff
@@ -85,20 +85,103 @@ module.exports = {
...
+              revisions: revisions
+            };
+          });
+      },
+
+      _list: function(/* context */) {
+        let bucket      = this.readConfig('bucket');
+        let deployArchive      = this.readConfig('deployArchive');
+        let revisionPrefix = deployArchive + "-";
+        let indexKey = deployArchive + '-info.json';
```
Comment (jeffjewiss) [inline index.js]: During testing I found that the name of the file uploaded to my bucket was 'fastboot-deploy-info.json' but my `indexKey` was `dist-info.json`. Can this be configurable or should it be updated?

Also, this causes an error if a `prefix` is used (ie. putting the files in a directory in the bucket). Could you please add the prefix value to the `indexKey` if it exists? This was done elsewhere with a ternary.

**Review by jeffjewiss — CHANGES_REQUESTED** [review]
@jheth thanks for this PR, this will be a great feature to have!

During testing I found one issue with the `_list` method and not handling prefix values. With some slight changes I was able to get it working and list out revisions for a project of mine. Nice work!

Otherwise, my comments are mostly style related. This would have been easier if better eslint rules were setup, sorry about that.

File: index.js [hunk 9]
```diff
@@ -85,20 +85,112 @@ module.exports = {
...
+      },
+
+      _list(/* context */) {
+        const bucket = this.readConfig('bucket');
+        const deployArchive = this.readConfig('deployArchive');
+        const archiveExt = `.${this.readConfig('archiveType')}`;
+        const deployInfo = this.readConfig('deployInfo');
+        const prefix = this.readConfig('prefix');
+        const archivePath = prefix ? [prefix, deployArchive].join('/') : deployArchive;
+        const indexKey = prefix ? [prefix, deployInfo].join('/') : deployInfo;
```
Comment (jheth) [inline index.js:130]: Thanks for pointing out the `prefix`, it also showed me that I was missing `deployInfo`.  In our deploy config we have `deployArchive` and `deployInfo` set to the same prefix so I missed the default `dist` scenario.

File: index.js (hunk 4 above)
Comment (jheth) [reply index.js:142]: It wasn't an issue in my testing.  If the file existed and started with a `{` then it parsed properly.

File: index.js [hunk 10]
```diff
@@ -85,20 +85,112 @@ module.exports = {
...
+              }
+            }
+          }
+
+          let results = data.revisions.Contents.sort(function(a, b) {
+            return new Date(b.LastModified) - new Date(a.LastModified);
+          }).map(function(d) {
+            let revision = '';
+            /* Check that this is the type of configured archive. */
+            if (d.Key.lastIndexOf(archiveExt) !== -1) {
```
Comment (jheth) [inline index.js:155]: I added this check to make sure the current file is an archive with the proper extension.

File: index.js [hunk 11]
```diff
@@ -85,20 +85,112 @@ module.exports = {
...
+              revision = d.Key.substring(revisionPrefix.length, d.Key.lastIndexOf('.'));
+            }
+            let active = data.current && revision === activeRevision;
+            return { revision: revision, timestamp: d.LastModified, active: active, deployer: 'fastboot-s3' };
+          }).filter(function(d) {
+            /* Filter out results where revision is empty. */
+            return d.revision !== '';
```
Comment (jheth) [inline index.js:162]: I also added this filter to return only those listings where we found a revision.  In our situation where archive and info files have the same prefix, it strips out the *-info.json file.

**Comment by jheth** [comment]
@jeffjewiss  This is ready for another review. Thanks for your time.

File: index.js (hunk 4 above)
Comment (jeffjewiss) [reply index.js:142]: Okay cool, thanks for clarifying this.

File: index.js (hunk 10 above)
Comment (jeffjewiss) [reply index.js:155]: Thanks for adding this.

File: index.js (hunk 9 above)
Comment (jeffjewiss) [reply index.js:130]: Ah okay cool, glad to hear this helped!

**Comment by jeffjewiss** [comment]
@jheth thanks for going through the review and making all of the changes.

This looks good to me and I was able to fetch revisions from this PR.

I don't have contributor access on NPM yet, so I won't be able to publish a new version with the updates right away. Hopefully @he9qi will have time to take a look or grant access, but I just wanted to give you a heads up.

**Comment by he9qi** [comment]
Looks good! 👍  @jheth @jeffjewiss .  Will publish a new version today!

**Comment by he9qi** [comment]
and @jeffjewiss just added you as a contributor on NPM!

**Comment by jeffjewiss** [comment]
@he9qi thanks for publishing a new version and adding me as a contributor. Appreciate it! 🙌

**Comment by kiwiupover** [comment]
@jheth good work here mate!

**Comment by he9qi** [comment]
@jeffjewiss np! thank you!
